### PR TITLE
CORE-4432 refresh e2e sessions periodically

### PR DESCRIPTION
**Pull request by emilybowe**
Changes 
---

Refresh e2e sessions periodically, so that session keys are rotated. 

At the point where sessions are created, they are now also scheduled for refresh. When the refresh occurs the heartbeat manager will also stop tracking the session at that point. A log will be generated.

Requires merge of: https://github.com/corda/corda-api/pull/636

Testing 
---

In addition to unit tests added this was tested with a manual deployment. The sending app simulator sent 24 messages of size 10000 bytes, in batches of 1, at 10 second intervals. This took 4 minutes, and refresh was triggered at the 2 minute mark. All of these messages were all received.

### CODE REVIEW

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 1]
```diff
@@ -132,17 +132,22 @@ internal class SessionManagerImpl(
     )
     private val outboundSessionPool = OutboundSessionPool(heartbeatManager::calculateWeightForSession)
 
+    private val fiveDaysInMilliseconds = 5*24*60*60*1000L
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]:
```
private val fiveDaysInMilliseconds = 5.days.toMillis()
```

Also, for testing, we should probably allow lower values.

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 2]
```diff
@@ -132,17 +132,22 @@ internal class SessionManagerImpl(
...
+
     private val publisher = PublisherWithDominoLogic(
         publisherFactory,
         coordinatorFactory,
         PublisherConfig(SESSION_MANAGER_CLIENT_ID, false),
         messagingConfiguration
     )
 
+    private val executorService = executorServiceFactory()
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:146]: We need to create it on `start`

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 3]
```diff
@@ -132,17 +132,22 @@ internal class SessionManagerImpl(
...
+
     override val dominoTile = ComplexDominoTile(
         this::class.java.simpleName,
         coordinatorFactory,
         ::onTileStart,
+        onClose = { executorService.shutdownNow() },
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:152]: If we create it on the object creation and close it on close, we can not restart it...

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 4]
```diff
@@ -545,9 +550,23 @@ internal class SessionManagerImpl(
             "Outbound session ${authenticatedSession.sessionId} established " +
                 "(local=${sessionCounterparties.ourId}, remote=${sessionCounterparties.counterpartyId})."
         )
+        executorService.schedule(
+            { refreshSessionAndLog(sessionCounterparties, message.header.sessionId) },
+            fiveDaysInMilliseconds,
+            TimeUnit.MILLISECONDS
```
Comment (yift-r3) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: Why milliseconds and not days (or minutes)?

Comment (emilybowe) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:556]: Millis seems consistent with the other scheduled threads

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 1 above)
Comment (williamvigorr3) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: We unit test this with the `MockTimeFacilitiesProvider` so don't need to adjust it for testing.

Comment (emilybowe) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: even if using mockTimeFacilitiesProvider.advanceTime ?

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 3 above)
Comment (williamvigorr3) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:152]: That is correct. It is not expected that a Lifecycle object can be restarted after it is closed (see KDOC on Lifecycle.close).

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 2 above)
Comment (williamvigorr3) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:146]: I don't think it really matters, when we create the executor service. I think it makes the code simplier to do it here rather than in `onStart` as we would then need to use an atomic reference.

We have moved away from the pattern of creating resources inside `onStart` and stopping them when the domino tile stops. Instead code needs to handle tasks running when the Lifecycle is Down.

File: components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt [hunk 5]
```diff
@@ -109,7 +111,7 @@ class P2PLayerEndToEndTest {
         .create(ConfigFactory.empty().withValue(INSTANCE_ID, ConfigValueFactory.fromAnyRef(1)))
 
     @Test
-    @Timeout(60)
+    @Timeout(600)
     fun `two hosts can exchange data messages over p2p using RSA keys`() {
```
Comment (emilybowe) [inline components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt]: fyi reverting this now

File: components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt [hunk 6]
```diff
@@ -152,9 +154,13 @@ class P2PLayerEndToEndTest {
...
-                hostBApplicationReaderWriter.close()
-                hostAMarkerReader.close()
+                hostAApplicationReader.stop()
+                hostBApplicationReaderWriter.stop()
+                hostAMarkerReader.stop()
+
+                val threeMinutesInMillis: Long = 3*60*1000
+                Thread.sleep(threeMinutesInMillis)
+
             }
```
Comment (emilybowe) [inline components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt]: fyi reverting this now

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 1 above)
Comment (yift-r3) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: Think about the QA that will need to wait for 5 days...

Comment (emilybowe) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:139]: one of the issues that had to be fixed only cropped up when it was set to the longer timeframe 5 days. So even if theoretically we made one for QA I don't know if that would truly test the behaviour.

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 4 above)
Comment (emilybowe) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: Changed to days now

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt [hunk 7]
```diff
@@ -136,17 +137,22 @@ internal class SessionManagerImpl(
     )
     private val outboundSessionPool = OutboundSessionPool(heartbeatManager::calculateWeightForSession)
 
+    private val fiveDays = 5.days.inWholeDays
```
Comment (dimosr) [inline components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: just curious; is there any benefit when compared to just doing:
```
private val fiveDays = 5L
```
I would also rename the variable to indicate what it represents e.g. `sessionRefreshThresholdInDays`

Comment (dimosr) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: although if I understood correctly, we might be refactoring this atm to allow any time unit for easy e2e testing. In that case, I suspect that will be changed anyway.

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 3 above)
Comment (dimosr) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:152]: I think we had discussed that in the past and we agreed it's fine to start these type of resources (thread pools) once and then stop them only once at application stop (not everytime the component goes up-and-down).

Although, I have to admit we are not doing that very precisely at the moment since the processors seem to invoke `stop()`, not `close()`. In practice, the thread pools will be forcibly stopped anyway on application stop, but something we should be looking at more widely when we get to clean up our domino logic a bit more. I think this is similar with the other usages of thread pools at the moment, so probably better to at least keep it consistent for now.

File: components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt [hunk 8]
```diff
@@ -382,7 +383,7 @@ class P2PLayerEndToEndTest {
         private val configPublisher = publisherFactory.createPublisher(PublisherConfig("config-writer", false), bootstrapConfig)
         private val gatewayConfig = createGatewayConfig(p2pPort, p2pAddress, sslConfig)
         private val tlsTenantId by lazy {
-            TLS_KEY_TENANT_ID
```
Comment (dimosr) [inline components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt]: What's the motivation behind these changes in this test? Is this an incorrectly-resolved conflict on merge from main?

These look similar to the refactorings I did recently on that test.

Comment (dimosr) [reply components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt]: Also, if all these changes are coming from the main branch (and they haven't been done in this PR explicitly), they shouldn't be showing up in this diff..

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt [hunk 9]
```diff
@@ -1610,4 +1612,125 @@ class SessionManagerTest {
...
+            SessionManager.SessionCounterparties(OUR_PARTY, PEER_PARTY)
+        )
+
+        verify(outboundSessionPool.constructed().last()).replaceSession(protocolInitiator.sessionId, protocolInitiator)
+        verify(publisherWithDominoLogicByClientId["session-manager"]!!.last())
+            .publish(listOf(Record(SESSION_OUT_PARTITIONS, protocolInitiator.sessionId, null))
+        )
+
+        loggingInterceptor.assertSingleWarningContains("No partitions from topic link.in are currently assigned " +
+                "to the inbound message processor. Sessions: [sessionId] will not be initiated."
```
Comment (dimosr) [inline components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: is this something relevant to this test case?

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt [hunk 10]
```diff
@@ -1610,4 +1612,125 @@ class SessionManagerTest {
...
+        )
+    }
+
+
+
+
+    @Test
+    fun `DRAFT`() {
```
Comment (dimosr) [inline components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: need to add proper test name

**Review by dimosr — COMMENTED** [review]
I had a quick look and left a few comments, but aware there are still some changes that are pending (making threshold configurable etc.)

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 1 above)
Comment (emilybowe) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt]: This will be made configurable so marking this resolved

File: components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt (hunk 8 above)
Comment (emilybowe) [reply components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt]: I did have an issue at the start where I had some local fixes to the e2e test so it would run like sleeping threads. Think I confused git by editing the merge file or not completing it properly. Then when I would fetch & merge or try pulling from main it wouldn't update and I had compile errors (and I tried invalidate cache and restart app and rebuilding - intellij wouldn't budge) so I asked Will and then made some manual updates and it compiled. I don't think the above were among them so maybe they got pulled in during the incomplete merge or something like that.

File: components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt (hunk 3 above)
Comment (emilybowe) [reply components/link-manager/src/main/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerImpl.kt:152]: ok so if I understand correctly - leave as is and this might be refactored with other thread pool usages when next cleaning up domino logic -> so marking resolved

File: components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt (hunk 8 above)
Comment (emilybowe) [reply components/link-manager/src/nonOsgiIntegrationTest/kotlin/net/corda/p2p/P2PLayerEndToEndTest.kt]: fyi Will explained the issue here - reverting these now

File: applications/tools/p2p-test/p2p-setup/src/main/kotlin/net/corda/p2p/setup/LinkManagerConfiguration.kt [hunk 11]
```diff
@@ -83,6 +83,12 @@ class LinkManagerConfiguration : Callable<Collection<Record<String, Configuratio
     )
     var sessionsPerPeer = 4L
 
+    @Option(
+        names = ["--sessionRefreshThreshold"],
+        description = ["Session refresh threshold in milliseconds - *important* do not change"]
```
Comment (dimosr) [inline applications/tools/p2p-test/p2p-setup/src/main/kotlin/net/corda/p2p/setup/LinkManagerConfiguration.kt]:
```suggestion
        description = ["Session refresh threshold in milliseconds"]
```
I'd remove the last part now given it's configurable.

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt [hunk 12]
```diff
@@ -1610,4 +1615,175 @@ class SessionManagerTest {
...
+        whenever(protocolInitiator.generateOurHandshakeMessage(eq(PEER_KEY.public), any())).thenReturn(mock())
+
+        val header = CommonHeader(MessageType.RESPONDER_HANDSHAKE, 1, protocolInitiator.sessionId, 4, Instant.now().toEpochMilli())
+        val responderHello = ResponderHelloMessage(header, ByteBuffer.wrap(PEER_KEY.public.encoded), ProtocolMode.AUTHENTICATED_ENCRYPTION)
+
+        sessionManager.processSessionMessage(LinkInMessage(responderHello))
+        val responderHandshakeMessage = ResponderHandshakeMessage(header, RANDOM_BYTES, RANDOM_BYTES)
+        val session = mock<Session>()
+
+        whenever(session.sessionId).doAnswer{protocolInitiator.sessionId}
```
Comment (dimosr) [inline components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: nit
```suggestion
        whenever(session.sessionId).thenReturn(protocolInitiator.sessionId)
```

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt [hunk 13]
```diff
@@ -1610,4 +1615,175 @@ class SessionManagerTest {
...
+        startSendingHeartbeats(sessionManager)
+
+        val heartbeatsExpected = 5
+
+        repeat(heartbeatsExpected) {
+            mockTimeFacilitiesProvider.advanceTime(longTimePeriodConfigWithHeartbeat.heartbeatPeriod.plus(5.millis))
+            sessionManager.messageAcknowledged(protocolInitiator.sessionId)
+        }
+
+        assertThat(heartbeatsExpected - messages.size).isEqualTo(1)
```
Comment (dimosr) [inline components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: Do I understand correctly that we expect 4 (instead of `heartbeatsExpected` = 5) messages, because in the last advancement of the time in the loop the session refresh threshold will be crossed and the session will be refreshed?

If that's the case, it would be good to add a comment here or maybe reduce the number of `heartbeatsExpected` to a lower value and then do an extra advancement of the time outside the loop to trigger the session to expire? Mostly to make it easier for whoever reads the code to understand what's going on.

**Review by dimosr — COMMENTED** [review]
Only some very minor comments - LGTM otherwise. I also left a small comment on the `corda-api` PR that might require an update here, if you follow my suggestion there.

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt (hunk 12 above)
Comment (emilybowe) [reply components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: I changed this, but have now changed it back. For some reason I get 'unfinished stubbing' test failure if I do it this suggested way (thenReturn), but passes if I do the original way (doAnswer)

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt (hunk 13 above)
Comment (emilybowe) [reply components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: I've made a change to address this - let me know what you think of the fix

**Review by williamvigorr3 — DISMISSED** [review]
LGTM

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt (hunk 12 above)
Comment (dimosr) [reply components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: The reason for this is Mockito doesn't like it when you invoke a mock object from within `thenReturn` (and `protocolInitiator` is a mock). So, doing something like:
```
whenever(session.sessionId).thenReturn("sessionId")
```
would remove the exception and the test would pass.

But, it's ok don't worry about for this, you can keep it as is. Just keep in mind that `thenReturn` is preferable when you want to just return a single value. And `doAnswer` when you want to execute more complicated logic to execute the return value.

File: components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt (hunk 13 above)
Comment (dimosr) [reply components/link-manager/src/test/kotlin/net/corda/p2p/linkmanager/sessions/SessionManagerTest.kt]: yup seems fine

**Review by dimosr — DISMISSED** [review]

**Review by williamvigorr3 — APPROVED** [review]
LGTM

**Review by dimosr — APPROVED** [review]
